frame/support/src/storage/stream_iter.rs

```diff
@@ -0,0 +1,405 @@
+/// the internal buffer is directly forwarded to the state to reduce the number of reads from the
+/// state.
+struct StorageInput {
+	key: Vec<u8>,
+	offset: u32,
+	total_length: u32,
+	exists: bool,
+	buffer: Vec<u8>,
+	buffer_pos: usize,
+}
+
+impl StorageInput {
+	/// Create a new instance of the input.
+	///
+	/// - `key`: The storage key of the storage item that this input will read.
+	fn new(key: Vec<u8>) -> Self {
+		let mut buffer = Vec::with_capacity(STORAGE_INPUT_BUFFER_CAPACITY);
+		unsafe {
+			buffer.set_len(buffer.capacity());
+		}
+
+		let (total_length, exists) =
+			if let Some(total_length) = sp_io::storage::read(&key, &mut buffer, 0) {
+				(total_length, true)
+			} else {
+				(0, false)
+			};
+
+		if (total_length as usize) < buffer.len() {
+			unsafe {
+				buffer.set_len(total_length as usize);
+			}
+		}
+
+		Self { total_length, offset: buffer.len() as u32, key, exists, buffer, buffer_pos: 0 }
+	}
+
+	/// Fill the internal buffer from the state.
+	fn fill_buffer(&mut self) -> Result<(), codec::Error> {
+		self.buffer.copy_within(self.buffer_pos.., 0);
+		let present_bytes = self.buffer.len() - self.buffer_pos;
+		self.buffer_pos = 0;
+
+		unsafe {
+			self.buffer.set_len(self.buffer.capacity());
+		}
+
+		if let Some(after_offset) =
+			sp_io::storage::read(&self.key, &mut self.buffer[present_bytes..], self.offset)
+		{
+			let buffer_len = if (after_offset as usize) < (self.buffer.len() - present_bytes) {
+				present_bytes + after_offset as usize
+			} else {
+				self.buffer.len()
+			};
+
+			unsafe {
+				self.buffer.set_len(buffer_len);
+			}
+
+			self.offset += (buffer_len - present_bytes) as u32;
+			self.buffer_pos = 0;
+
+			Ok(())
+		} else {
+			// The value was deleted, let's ensure we don't read anymore.
+			self.offset = self.total_length;
+			unsafe {
+				self.buffer.set_len(0);
+			}
+
+			Err("Value doesn't exist in the state?".into())
+		}
+	}
+
+	/// Returns if the value to read exists in the state.
+	fn exists(&self) -> bool {
+		self.exists
+	}
+}
+
+impl codec::Input for StorageInput {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		Ok(Some(self.total_length.saturating_sub(self.buffer_pos as u32) as usize))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		if into.len() > self.buffer.capacity() {
+			let num_cached = self.buffer.len() - self.buffer_pos;
+
+			into[..num_cached].copy_from_slice(&self.buffer[self.buffer_pos..]);
+			self.buffer_pos = self.buffer.len();
+
+			if let Some(after_offset) =
+				sp_io::storage::read(&self.key, &mut into[num_cached..], self.offset)
+			{
+				let required_to_read = (into.len() - num_cached) as u32;
+
+				if after_offset < required_to_read {
+					return Err("not enough data to fill the buffer".into())
+				}
+
+				self.offset += required_to_read;
+				return Ok(())
+			} else {
+				return Err("Value doesn't exist in the state?".into())
+			}
```

Comment:
Inside of `fill_buffer` we set `self.offset` to `total_length` in case `read` returns `None`, but we don't do that here. We should probably be consistent and do the same thing in both cases.